File: justtakethestairs/submissions/time_limit_exceeded/brute_force.py
#The basic idea of the brute force solution is:
#"Try every possible way to split the groups into trips, take the minimum cost".

N, C = map(int, input().split())
S = [0] * (N + 1)
F = [0] * (N + 1)
for i in range(1, N + 1):
    S[i], F[i] = map(int, input().split())

best = float("inf")
stack = [(1, 0)]
while stack:
    start, cost = stack.pop()
    if start > N:
        best = min(best, cost)
        continue
    students = 0
    floors = set()
    for end in range(start, N + 1):
        students += S[end]
        if students > C:
            break
        floors.add(F[end])
        max_floor = max(floors)
        stops = len(floors)
        trip_cost = max_floor + stops + max_floor
        if end == N:
            trip_cost -= max_floor
        stack.append((end + 1, cost + trip_cost))
print(best)

File: justtakethestairs/submissions/time_limit_exceeded/brute_force.rs
use std::io::{self, BufRead, BufReader};
use std::cmp;
use std::collections::HashSet;

fn main() {
    let stdin = io::stdin();
    let mut lines = BufReader::new(stdin.lock()).lines();
    let line = lines.next().unwrap().unwrap();
    let mut iter = line.split_whitespace();
    let N: usize = iter.next().unwrap().parse().unwrap();
    let C: i64 = iter.next().unwrap().parse().unwrap();
    let mut S = vec![0i64; N+1];
    let mut F = vec![0i64; N+1];
    for i in 1..=N {
        let line = lines.next().unwrap().unwrap();
        let mut iter = line.split_whitespace();
        S[i] = iter.next().unwrap().parse().unwrap();
        F[i] = iter.next().unwrap().parse().unwrap();
    }

    let mut best = 1_000_000_000;
    let mut stack = vec![(1, 0)];
    while stack.len() != 0 {
        let (start, cost) = stack.pop().unwrap();
        if start > N {
            best = cmp::min(best, cost);
            continue;
        }
        let mut students = 0;
        let mut floors: HashSet<i64> = HashSet::new();
        for end in start..=N {
            students += S[end];
            if students > C {
                break;
            }
            floors.insert(F[end]);
            let max_floor = *floors.iter().max().unwrap();
            let stops = floors.len() as i64;
            let mut trip_cost = max_floor + stops + max_floor;
            if end == N {
                trip_cost -= max_floor;
            }
            stack.push((end + 1, cost + trip_cost))
        }
    }
    println!("{}", best);
}